### PR TITLE
Fix (Read,Invoke) denied default provider handling

### LINKED ISSUE #9055: aws.GetAvailabilityZones fails when the default aws provider is disabled
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

This seems to be a regression in 3.25.0 (which includes a bug fix to allow disabling of the `kubernetes` default when using `helmv3`). Specifically, when the default `aws` provider is disabled and you use `aws.GetAvailabilityZones`, you get the following error:

```
    	* rpc error: code = Unknown desc = Default provider for 'aws' disabled. 'aws:index/getAvailabilityZones:getAvailabilityZones' must use an explicit provider.
    exit status 1
```

This works fine in 3.24.1

### Steps to reproduce

Here is a Go program that shows this:

```
package main

import (
	"fmt"

	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"

	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		awsProvider, err := aws.NewProvider(ctx, "r53-aws-provider", &aws.ProviderArgs{
			SkipMetadataApiCheck: pulumi.Bool(false),
		})
		if err != nil {
			return err
		}

		state := "available"
		availableAZ, err := aws.GetAvailabilityZones(ctx, &aws.GetAvailabilityZonesArgs{
			State: &state,
		}, pulumi.Provider(awsProvider))
		if err != nil {
			fmt.Println("ERROR")
			return err
		}

		fmt.Println(availableAZ.Names)

		return nil
	})
}
```

with go.mod:
```
module pulumi_test

go 1.16

require (
	github.com/pulumi/pulumi-aws/sdk/v4 v4.37.5 // indirect
	github.com/pulumi/pulumi/sdk/v3 v3.25.0
)
```

and versions:

```
$ pulumi version
v3.25.0
$ pulumi plugin ls
NAME        KIND      VERSION  SIZE    INSTALLED      LAST USED
aws         resource  4.37.5   350 MB  9 minutes ago  now
aws         resource  4.17.0   276 MB  6 months ago   5 days ago
aws         resource  4.7.0    267 MB  6 months ago   5 days ago
aws         resource  4.6.0    261 MB  3 weeks ago    5 days ago
aws         resource  4.5.1    260 MB  9 months ago   5 days ago
aws         resource  4.0.0    255 MB  9 months ago   1 week ago
aws         resource  1.31.0   222 MB  9 months ago   1 week ago
command     resource  0.0.3    23 MB   1 month ago    6 days ago
kubernetes  resource  3.6.3    73 MB   6 months ago   5 days ago
kubernetes  resource  3.3.1    72 MB   6 months ago   1 week ago
kubernetes  resource  3.2.0    72 MB   9 months ago   5 days ago
kubernetes  resource  3.1.1    72 MB   9 months ago   5 days ago
postgresql  resource  3.0.1    39 MB   9 months ago   5 days ago
random      resource  4.3.1    32 MB   2 days ago     2 days ago
random      resource  4.1.1    30 MB   9 months ago   5 days ago
random      resource  1.7.0    38 MB   9 months ago   1 week ago
tls         resource  4.0.0    30 MB   9 months ago   5 days ago
```

Expected: this should work
Actual: it doesn't work, throws the above error

### LINKED ISSUE #9060: Pulumi 3.25.0 breaks pulumi_aws.iam.get_policy_document with explicit AWS provider
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

Pulumi 3.25.0
Runtime: python
Pulumi AWS: pulumi_aws-4.37.5

```
$ cat Pulumi.prod.yaml
config:
  pulumi:disable-default-providers:
    - aws

```

And I have the code like the following (simplified):

```
aws_provider = ...
print(aws_provider)
policy = pulumi_aws.iam.get_policy_document(
    statements=[
        pulumi_aws.iam.GetPolicyDocumentStatementArgs(
            actions=['s3:*'],
        ),
    ],
    opts=pulumi.InvokeOptions(provider=aws_provider),
)

```

`print(aws_provider)` result:
```
<pulumi_aws.provider.Provider object at 0x7f559c72d940>
```

and an error is raised:
```
Exception: invoke of aws:iam/getPolicyDocument:getPolicyDocument failed: Default provider for 'aws' disabled. 'aws:iam/getPolicyDocument:getPolicyDocument' must use an explicit provider.
```

When I just change the Pulumi version to 3.24.1 (with no other changes in the code), the issue disappears.

### PR DESCRIPTION
**Pull request by iwahbe**
When denying default providers was added, we had no special handling for
Reads and Invokes. This lead to confusing error messages. The fix (#8853)
involved checking on invokes. This check didn't apply to several types
of calls (Read) as well as blocking invokes with providers applied.

This PR fixes the logic to only deny providers when they are default
providers.

It also pushes the change into `getProviderFromSource`, which ensures
that this behavior is handled the same way (and correctly) for both
Invokes and Reads.

# Description

Fixes #9055
Fixes #9060

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by AaronFriel** [comment]
Confirm that this fixes #9060 as well.

File: pkg/resource/deploy/source_eval.go [hunk 1]
```diff
@@ -576,16 +576,21 @@ func getProviderReference(defaultProviders *defaultProviders, req providers.Prov
...
+	req providers.ProviderRequest, rawProviderRef string,
+	token tokens.ModuleMember) (plugin.Provider, error) {
 
 	providerRef, err := getProviderReference(defaultProviders, req, rawProviderRef)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("getProviderFromSource: %w", err)
+	} else if providers.IsDenyDefaultsProvider(providerRef) {
+		msg := diag.GetDefaultProviderDenied("Invoke").Message
+		return nil, fmt.Errorf(msg, req.Package(), token)
```
Comment (AaronFriel) [inline pkg/resource/deploy/source_eval.go:588]: Nit: Does "req.Package()" return a string containing a format tag? If so, code smell?

Either way, would it be easy to refactor & remove token as an argument to the function? It's only used in the error path and callers could insert it themselves.

File: pkg/resource/deploy/source_eval.go [hunk 2]
```diff
@@ -850,6 +849,9 @@ func (rm *resmon) ReadResource(ctx context.Context,
 		ref, provErr := rm.defaultProviders.getDefaultProviderRef(providerReq)
 		if provErr != nil {
 			return nil, provErr
+		} else if providers.IsDenyDefaultsProvider(ref) {
+			msg := diag.GetDefaultProviderDenied("Read").Message
```
Comment (AaronFriel) [inline pkg/resource/deploy/source_eval.go:853]: I believe this function takes a urn as argument, do we have one here?

File: pkg/resource/deploy/source_eval_test.go [hunk 3]
```diff
@@ -117,6 +125,27 @@ func newProviderEvent(pkg, name string, inputs resource.PropertyMap, parent reso
...
+func disableDefaultProviders(runInfo *EvalRunInfo, pkgs ...string) {
+	if runInfo.Target.Config == nil {
+		runInfo.Target.Config = config.Map{}
+	}
+	c := runInfo.Target.Config
+	key := config.MustMakeKey("pulumi", "disable-default-providers")
+	if _, ok, err := c.Get(key, false); err != nil {
+		panic(err)
+	} else if ok {
+		panic("disbaleDefaultProviders cannot be called twice")
```
Comment (AaronFriel) [inline pkg/resource/deploy/source_eval_test.go:137]: sp

File: pkg/resource/deploy/source_eval_test.go [hunk 4]
```diff
@@ -492,6 +521,168 @@ func TestReadInvokeDefaultProviders(t *testing.T) {
 	assert.Equal(t, expectedInvokes, int(invokes))
 }
 
+// Test that we can run operations with default providers disabled.
+//
+// We run against the matrix of
+// - enabled  vs disabled
+// - explicit vs default
+//
+// B exists as a sanity check, to ensure that we can still do perform arbitrary
```
Comment (AaronFriel) [inline pkg/resource/deploy/source_eval_test.go:530]: nit: "do perform"

File: pkg/resource/deploy/source_eval_test.go [hunk 5]
```diff
@@ -492,6 +521,168 @@ func TestReadInvokeDefaultProviders(t *testing.T) {
...
+		hasExplicit    bool
+		expectFail     bool
+	}
+	cases := []TT{}
+	for _, disableDefault := range []bool{true, false} {
+		for _, hasExplicit := range []bool{true, false} {
+			cases = append(cases, TT{
+				disableDefault: disableDefault,
+				hasExplicit:    hasExplicit,
+				expectFail:     disableDefault && !hasExplicit,
```
Comment (AaronFriel) [inline pkg/resource/deploy/source_eval_test.go:544]: 👍 surprisingly easy to read test matrix thanks to this conditional

**Review by AaronFriel — APPROVED** [review]
LGTM modulo nits. I'd like to see "token" removed from the arguments to `getProviderFromSource` unless there's a need for it.

File: pkg/resource/deploy/source_eval.go (hunk 1 above)
Comment (iwahbe) [reply pkg/resource/deploy/source_eval.go:588]: `req.Package()` does not contain a format tag, but `msg` contains 2. I'm not sure how we would refactor it out without changing the calling convention of our `diag` error messages.

File: pkg/resource/deploy/source_eval.go (hunk 2 above)
Comment (iwahbe) [reply pkg/resource/deploy/source_eval.go:853]: The urn is generated by the provider after the read. So we don't have access to it yet.

File: pkg/resource/deploy/source_eval.go (hunk 1 above)
Comment (AaronFriel) [reply pkg/resource/deploy/source_eval.go:588]: got it. seems out of scope for this PR.

**Pull request merged**